### PR TITLE
FC-963 Add tspo index (and index query refactor)

### PR DESCRIPTION
**Pull request by zonotope**
This patch adds and maintains a new `t s p o` sorted index to ledgers, but this index is not used in any queries just yet.

The meat of the changes however are in the `fluree.db.query.range` namespace (which I believe will be collapsed by default in this pull request because the change is so large, so please make sure to review that file explicitly).

This refactor changes the way we use core.async channels to make them behave more like lazy sequences that process index nodes and flakes one-by-one. I believe this allows the code to be much more accessible, maintainable, and streamlined while also allowing for more efficient processing as each node is only processed as necessary.

After this patch is accepted, the remaining tasks are to use the new tspo index in block-related queries and reconstruct our previous block formatted data from the tspo index and novelty.

### CODE REVIEW

File: src/fluree/db/query/range.cljc [hunk 1]
```diff
@@ -56,154 +50,234 @@
...
+     (go
+       (let [start-flake (<? (resolve-match-flake db start-test start-parts))
+             end-flake   (<? (resolve-match-flake db end-test end-parts))]
+         (-> db
+             (index-node-stream idx start-flake end-flake)
+             (expand-history-range from-t to-t novelty start-test start-flake
+                                   end-test end-flake)
+             (filter-authorized db start-flake end-flake)
+             (take-only limit)
+             (->> (async/into #{}))
```
Comment (bplatz) [inline src/fluree/db/query/range.cljc:221]: I think out-chan is already going to put into a set, might be more efficient to put into a list or vector. Set placement will require individual equality checks - which get done twice here.

File: src/fluree/db/query/range.cljc [hunk 2]
```diff
@@ -56,154 +50,234 @@
...
+  (let [flakes-xf   (map :flakes)
+        subrange-xf (flake-subrange-xf start-test start-flake end-test end-flake)
+        xforms      (cond-> [flakes-xf subrange-xf]
+                      subject-fn   (conj (filter (fn [^Flake f]
+                                                   (subject-fn (.-s f)))))
+                      predicate-fn (conj (filter (fn [^Flake f]
+                                                   (predicate-fn (.-p f)))))
+                      object-fn    (conj (filter (fn [^Flake f]
+                                                   (object-fn (.-o f))))))]
+    (apply comp xforms)))
```
Comment (bplatz) [inline src/fluree/db/query/range.cljc:259]: 🥇

File: src/fluree/db/query/range.cljc [hunk 3]
```diff
@@ -220,120 +294,54 @@
...
+                                   :object-fn object-fn
+                                   :start-test start-test
+                                   :start-flake start-flake
+                                   :end-test end-test
+                                   :end-flake end-flake})
+            (filter-authorized db start-flake end-flake)
+            (select-subject-window {:subject-limit subject-limit
+                                    :flake-limit flake-limit
+                                    :offset offset})
+            (->> (async/into #{}))
```
Comment (bplatz) [inline src/fluree/db/query/range.cljc:342]: Same comment as time index range above... two redundant set operations (I think)

**Review by bplatz — COMMENTED** [review]
Some general comments, may not all be applicable.

File: src/fluree/db/query/range.cljc [hunk 4]
```diff
@@ -56,154 +50,234 @@
...
+        (async/close! subrange-ch)))
+    subrange-ch))
+
+(defn filter-authorized
+  "Returns a channel that will eventually contain only the schema flakes and the
+  flakes validated by fluree.db.permissions-validate/allow-flake? function for
+  the database `db` from the `flake-stream` channel"
+  [flake-stream {:keys [permissions] :as db} ^Flake start ^Flake end]
+  #?(:cljs
+     flake-stream ; Note this bypasses all permissions in CLJS for now!
```
[the remark this answers is not in the document]
Comment (zonotope) [reply src/fluree/db/query/range.cljc:137]: It looks like the permissions are unchecked for all cljs code [here](https://github.com/fluree/db/blob/master/src/fluree/db/query/range.cljc#L132-L133) and [here](https://github.com/fluree/db/blob/master/src/fluree/db/query/range.cljc#L154-L155)

**Review by bplatz — APPROVED** [review]

**Pull request merged**